src/v/transform/logging/log_manager.cc

```diff
@@ -0,0 +1,297 @@
+ss::future<> manager<ClockType>::stop() {
+    _as.request_abort();
+    co_await _flusher->stop();
+    if (!_gate.is_closed()) {
+        co_await _gate.close();
+    }
+}
+
+template<typename ClockType>
+bool manager<ClockType>::check_lwm() const {
+    return _buffer_sem.available_units() <= _buffer_low_water_mark;
+}
+
+template<typename ClockType>
+void manager<ClockType>::enqueue_log(
+  ss::log_level level,
+  model::transform_name_view transform_name,
+  std::string_view message) {
+    auto msg_len =
+      [lim = _line_limit_bytes()](std::string_view message) -> size_t {
+        return std::min(lim, message.size());
+    };
+
+    auto validate_msg =
+      [&msg_len](std::string_view message) -> std::optional<ss::sstring> {
+        auto sub_view = message.substr(0, msg_len(message));
+        if (!is_valid_utf8(sub_view)) {
+            return std::nullopt;
+        } else if (contains_control_character(sub_view)) {
+            // escape control chars and truncate (again, if necessary)
+            auto res = replace_control_chars_in_string(sub_view);
+            return res.substr(0, msg_len(res));
+        }
+        return ss::sstring{sub_view.data(), sub_view.size()};
+    };
+
+    auto get_queue = [this](std::string_view name) {
+        auto res = _log_buffers.find(name);
+        if (res == _log_buffers.end()) {
+            auto [it, ins] = _log_buffers.emplace(
+              ss::sstring{name.data(), name.size()}, buffer_t{});
+            if (ins) {
+                return it;
+            }
+            // otherwise something went badly wrong. we'll return the end
+            // iterator and fail the operation
+            vlog(
+              tlg_log.error, "Failed to enqueue transform log: Buffer error");
+        }
+        return res;
+    };
+
+    auto it = get_queue(transform_name);
+    if (it == _log_buffers.end()) {
+        return;
+    }
+
+    // Unfortunately, we don't know how long an escaped string will be
+    // until we've allocated memory for it. So we optimistically grab
+    // units for the unmodified log message here, hoping that, in the
+    // common case, no control chars are present. If the message fails
+    // validation, we will simply return the units to _buffer_sem on
+    // function return.
+    // NOTE(oren): we can still truncate a view up front w/o allocating
+    message = message.substr(0, msg_len(message));
+    auto units = ss::try_get_units(_buffer_sem, message.size());
+    if (!units) {
+        vlog(tlg_log.debug, "Failed to enqueue transform log: Buffer full");
+        return;
+    }
+
+    auto b = validate_msg(message);
+    if (!b.has_value()) {
+        vlog(
+          tlg_log.debug,
+          "Failed to enqueue transform log: Message validation failed");
+        return;
+    }
+
+    it->second.emplace_back(
+      event{_self, event::clock_type::now(), level, std::move(*b)},
+      std::move(*units));
+
+    if (check_lwm() && _flusher != nullptr) {
```

Comment:
When can `_flusher` be `nullptr`?

Reply:
If `enqueue_log` were called before `manager::start`. I think this shouldn't happen, but without all the wiring in place I'm not 100% sure. Maybe an assertion would express this better.

Reply:
Probably? I know for things that initialize metrics there's a good reason to construct in `start` (global side effects?). Not an issue here, but I've seen this pattern on a couple other classes. TBH in this case I'm just cargo-culting a bit.

Reply:
Oh, yeah transform::service does this because other dependent sharded services are not initialized yet - kind of annoying. I think in this case, we don't have this concern because the memory is owned by the manager, so we can simplify.